tools/sync.js

```diff
@@ -0,0 +1,214 @@
+              console.log('Skip: Duplicate key ' +   blockData.number.toString() + ': ' + err);
+            }
+        } else {
+           console.log('Error: Aborted due to error on ' + 'block number ' + blockData.number.toString() + ': ' +  err);
+           process.exit(9);
+       }
+    } else {
+        if(!('quiet' in config && config.quiet === true)) {
+            console.log('DB successfully written for block number ' + blockData.number.toString());
+        }
+    }
+  });
+}
+/**
+    Break transactions out of blocks and write to DB
+**/
+var writeTransactionsToDB = function(config, blockData) {
+  var bulkOps = [];
+  if (blockData.transactions.length > 0) {
+      for (d in blockData.transactions) {
+          var txData = blockData.transactions[d];
+          txData.timestamp = blockData.timestamp;
+          txData.value = etherUnits.toEther(new BigNumber(txData.value), 'wei');
+          bulkOps.push(txData);
+      }
+      Transaction.collection.insert(bulkOps, function( err, tx ){
+          if ( typeof err !== 'undefined' && err ) {
+              if (err.code == 11000) {
+                  console.log('Skip: Duplicate key ' +
+                  err);
+              } else {
+                 console.log('Error: Aborted due to error: ' +
+                      err);
+                 process.exit(9);
+             }
+          } else if(!('quiet' in config && config.quiet === true)) {
+              console.log(blockData.transactions.length.toString() + ' transactions recorded for Block# ' + blockData.number.toString());
+          }
+      });
+  }
+}
+var checkBlockDBExistsThenWrite = function(config, blockData) {
+  Block.find({number: blockData.number}, function (err, b) {
+      if (!b.length){
+          writeBlockToDB(config, blockData);
+          writeTransactionsToDB(config, blockData);
+      }else if(!('quiet' in config && config.quiet === true)) {
+          console.log('Block number: ' + blockData.number.toString() + ' already exists in DB.');
+          listenBlocks(config);
+      }
+  });
+};
+var updatedEndBlock = function(config,lastBlock){
+  var configFile = '../conf.json';
+  var file = require(configFile);
+
+  file.endBlock = lastBlock;
+
+  fs.writeFile('conf.json', JSON.stringify(file, null, 2), function (err) {
+    if (err) return console.log(err);
+    //console.log('Wirting new Synced Block ' + lastBlock + ' to ' + configFile);
+  });
+};
+/**
+Take the last block the grabber exited on and update the param 'end' in the config.JSON
+**/
+var updateLastSynced = function(config,lastSync){
+  var configFile = '../conf.json';
+  var file = require(configFile);
+
+  file.lastSynced = lastSync;
+  config.lastSynced = lastSync;
+
+  fs.writeFile('conf.json', JSON.stringify(file, null, 2), function (err) {
+    if (err) return console.log(err);
+    //console.log('writing block ' + lastSync + ' to ' + configFile);
+
+    if (config.lastSynced === config.startBlock){
```

Comment:
Logic regarding `lastSynced` being here, and in `grabBlock2` is likely leading to some issues.

`updateLastSynced` should be the only function that deals with this. And its behavior should be agnostic from whether or not we're starting a new sync or an old one. Block 0 isn't special.

Reply:
What this is for, is when you set the flag `syncAll` to true in conf.json. The sync will keep on syncing till it reaches the value set in `startBlock`.

In theory this block explorer should work with out any issues on any GETH client. With more multi GETH clients being developed and not ever chain will want to sync all the way to block 0.

This makes this explorer more versatile.